modules/account/src/main/kotlin/edu/stanford/spezi/module/account/login/LoginViewModel.kt

```diff
@@ -85,37 +113,52 @@ class LoginViewModel @Inject constructor(
         navigator.navigateTo(
             AccountNavigationEvent.RegisterScreen(
                 isGoogleSignUp = false,
-                email = uiState.value.email,
-                password = uiState.value.password,
+                email = uiState.value.email.value,
+                password = uiState.value.password.value,
             )
         )
     }
 
     private fun updateTextField(
         action: Action.TextFieldUpdate,
-        it: UiState,
+        uiState: UiState,
     ): UiState {
-        val newValue = action.newValue
+        val newValue = FieldState(action.newValue)
+        val result = when (action.type) {
+            TextFieldType.PASSWORD -> validator.isValidPassword(action.newValue)
+            TextFieldType.EMAIL -> validator.isValidEmail(action.newValue)
+        }
+        val error =
+            if (hasAttemptedSubmit && result is FormValidator.Result.Invalid) result.errorMessageOrNull() else null
         return when (action.type) {
-            TextFieldType.PASSWORD -> it.copy(password = newValue)
-            TextFieldType.EMAIL -> it.copy(email = newValue)
+            TextFieldType.PASSWORD -> uiState.copy(
+                password = newValue.copy(error = error),
+                isFormValid = validator.isFormValid(uiState),
+                isPasswordSignInEnabled = uiState.email.value.isNotEmpty() && newValue.value.isNotEmpty()
+            )
+
+            TextFieldType.EMAIL -> uiState.copy(
+                email = newValue.copy(error = error),
+                isFormValid = validator.isFormValid(uiState),
+                isPasswordSignInEnabled = newValue.value.isNotEmpty() && uiState.password.value.isNotEmpty()
+            )
         }
     }
 
     private fun forgotPassword() {
-        if (uiState.value.email.isEmpty()) {
-            messageNotifier.notify("Please enter your email")
+        if (validator.isValidEmail(uiState.value.email.value) is FormValidator.Result.Valid) {
```

Comment:
Can we update all usages of `is Result.Valid/Invalid` with `.isValid/.isValid.not()` as well? A simple string would show them all :)